### ISSUE TITLE
Calling dismiss dependency fails when executed from a secondary effect (navigation beta)

### ISSUE THREAD

**Issue by scogeo** [pointfreeco/swift-composable-architecture#1947]
### Description

In the following example, the `dismiss` dependency fails to dismiss the sheet when it is called from an effect that has been sent back into the store (for example, after saving state changes to disk).

```swift
import SwiftUI
import ComposableArchitecture

struct AppReducer: ReducerProtocol {
    
    struct State: Equatable {
        @PresentationState var sheet: MySheet.State?
    }
 
    enum Action: Equatable {
        case sheet(PresentationAction<MySheet.Action>)
        case sheetButtonTapped
    }
    
    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .sheetButtonTapped:
                state.sheet = .init()
                return .none
                
            case .sheet:
                return .none
            }
        }
        .ifLet(\.$sheet, action: /Action.sheet) {
            MySheet()
        }
    }
}

struct MySheet: ReducerProtocol {
    
    struct State: Equatable {
    }
    
    enum Action: Equatable {
        case cancelTapped
        case saveTapped
        case saveCompleted
    }
    
    @Dependency(\.dismiss) var dismiss
    
    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch (action) {
            case .cancelTapped:
                return .fireAndForget { await self.dismiss() }
        
            case .saveCompleted:
                // This has no effect
                return .fireAndForget { await self.dismiss() }
                
            case .saveTapped:
                // Simulate saving data
                return .task {
                    try await Task.sleep(nanoseconds: 100 * NSEC_PER_MSEC)
                    return .saveCompleted
                }
            }
        }
    }
    
}

struct ContentView: View {
    let store: StoreOf<AppReducer>
    
    var body: some View {
        WithViewStore(store, observe: { $0} ) { viewStore in
            Button("Open sheet") { viewStore.send(.sheetButtonTapped) }
                .sheet(store: store.scope(state: \.$sheet, action: AppReducer.Action.sheet)) { store in
                    SheetView(store: store)
                }
        }
    }
}

struct SheetView: View {
    let store: StoreOf<MySheet>
    
    var body: some View {
        WithViewStore(store, observe: { $0} ) { viewStore in
            NavigationStack {
                Text("A Sheet")
                .toolbar {
                    ToolbarItem(placement: .cancellationAction) {
                        Button("Cancel") {
                            viewStore.send(.cancelTapped)
                        }
                    }
                    ToolbarItem(placement: .primaryAction) {
                        Button("Save") {
                            viewStore.send(.saveTapped)
                        }
                    }
                }
            }
            
        }
        
    }
}

struct ContentView_Previews: PreviewProvider {
    static var previews: some View {
        ContentView(store: .init(initialState: .init(), reducer: AppReducer()))
    }
}
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The sheet should be dismissed via an asynchronous effect.

### Actual behavior

The sheet is not dismissed via an asynchronous effect.

### Steps to reproduce

Copy the sample code into Xcode and run the preview.

### The Composable Architecture version information

navigation-beta (45b88b6f1b08c10cbd6141d9794828cb0a3f012a)

### Destination operating system

iOS 16.2

### Xcode version information

14.2

**Comment by mbrandonw** [comment]
Hey @scogeo, thanks so much for the clear, reproducible code sample! We see the problem and have a fix (#1948) that we will probably merge soon, but we have a few things we want to understand about the whole system first.

**Comment by scogeo** [comment]
@mbrandonw glad you were able to reproduce. Thanks to @stephencelis for jumping on the PR almost immediately last night! I read the PR comments and it looks like it uncovered a few additional issues so happy to wait for you guys to get it right.

I can hold off on deleting all of my delegated dismiss actions in my codebase for a few days more. 😀
